crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -468,6 +874,320 @@ fn derive_label(input: DeriveInput, label_type: Ident) -> TokenStream2 {
     }
 }
 
+struct FetchImplTokens<'a> {
+    struct_name: Ident,
+    fetch_struct_name: Ident,
+    state_struct_name: Ident,
+    fetch_trait_punctuated_lifetimes: Punctuated<GenericParam, Token![,]>,
+    impl_generics: ImplGenerics<'a>,
+    ty_generics: TypeGenerics<'a>,
+    where_clause: Option<&'a WhereClause>,
+    struct_has_world_lt: bool,
+    world_lt: GenericParam,
+    state_lt: GenericParam,
+}
+
+fn fetch_impl_tokens(ast: &DeriveInput) -> FetchImplTokens {
+    let world_lt = ast.generics.params.first().and_then(|param| match param {
+        lt @ GenericParam::Lifetime(_) => Some(lt.clone()),
+        _ => None,
+    });
+    let struct_has_world_lt = world_lt.is_some();
+    let world_lt = world_lt.unwrap_or_else(|| {
+        GenericParam::Lifetime(LifetimeDef::new(Lifetime::new("'world", Span::call_site())))
+    });
+    let state_lt =
+        GenericParam::Lifetime(LifetimeDef::new(Lifetime::new("'state", Span::call_site())));
+
+    let mut fetch_trait_punctuated_lifetimes = Punctuated::<_, Token![,]>::new();
+    fetch_trait_punctuated_lifetimes.push(world_lt.clone());
+    fetch_trait_punctuated_lifetimes.push(state_lt.clone());
+
+    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+
+    let struct_name = ast.ident.clone();
+    let fetch_struct_name = Ident::new(&format!("{}Fetch", struct_name), Span::call_site());
+    let state_struct_name = Ident::new(&format!("{}State", struct_name), Span::call_site());
+
+    FetchImplTokens {
+        struct_name,
+        fetch_struct_name,
+        state_struct_name,
+        fetch_trait_punctuated_lifetimes,
+        impl_generics,
+        ty_generics,
+        where_clause,
+        struct_has_world_lt,
+        world_lt,
+        state_lt,
+    }
+}
+
+struct WorldQueryFieldTypeInfo {
+    /// We convert `Mut<T>` to `&mut T` (because this is the type that implements `WorldQuery`)
+    /// and store it here.
+    query_type: Type,
+    /// The same as `query_type` but with `'fetch` lifetime.
+    fetch_init_type: Type,
+    is_read_only: bool,
+    is_phantom: bool,
+    readonly_types_to_assert: Vec<TypePath>,
```

Comment:
```suggestion
    read_only_types_to_assert: Vec<TypePath>,
```
Style consistency nit to match `is_read_only`.

Reply:
Ah, I guess I'll rename `is_read_only` to `is_readonly`... Or we should probably change the attribute to `#[read_only]` then. Which do you think is best?

Reply:
IMO stick with underscores consistently. It's easier to read and more logical (they're separate words), especially at a glance.

Reply:
Hm, I've found some other examples in Bevy code, where a variable is spelt as `readonly`. Also, there is an HTML attribute called `readonly`, GLSL also uses `readonly` as an annotation. So I'm leaning towards this variant.

I've renamed `is_read_only` to `is_readonly`. But I can still revert this in case we'll find `read_only` to be preferable.

Reply:
Works for me. We can do a proper consistency pass in a later PR if we care enough.